swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/TunnelArchiveByteStream.swift

```
@@ -0,0 +1,146 @@
+///
+/// Since the IPC channel is unidirectional from app -> tunnel, we use a simple data format to pass chunks
+/// of this memory buffer from the tunnel back to the app, including a boolean `done` to indicate when the
+/// app should stop reading from this buffer. The buffer is flushed when it reaches a configurable limit.
+///
+/// Currently this limit is set to 1 MB (chosen somewhat arbitrarily based on limited information found on the
+/// web), but can be easily enlarged in the future to reduce the number of IPC calls required to consume
+/// the entire archive. The LZFSE compression algorithm used by default in the Apple Archive Framework is
+/// quite efficient -- compression ratios for our logs can be as high as 100:1 using this format.
+///
+/// To prevent the tunnel from writing to this buffer faster than the app can consume it, we implement a basic
+/// backpressure mechanism that sleeps the tunnel's log export thread if the buffer is full until the app makes
+/// another IPC call to wake it up. The buffer is "flushed" either when the data is fully consumed (i.e. `close()` is
+/// called) or when the buffer size limit is hit. Only then will the app's completionHandler callback be called.
+public class TunnelArchiveByteStream: ArchiveByteStreamProtocol {
+  struct Chunk: Codable {
+    var done: Bool = false
+    var data = Data()
+  }
+
+  let chunkSize = 1024 * 1024 // 1 MB
+  let encoder = PropertyListEncoder()
+  var chunk: Chunk
+  var chunkHandler: ((Data?) -> Void)?
+  var logger: Log
+  var completionHandler: () -> Void
+
+  public init(
+    logger: Log,
+    chunkHandler: @escaping (Data?) -> Void,
+    completionHandler: @escaping () -> Void
+  ) {
+    self.completionHandler = completionHandler
+    self.logger = logger
+    self.chunk = Chunk()
+    self.chunkHandler = chunkHandler
+  }
+
+  public func read(into buffer: UnsafeMutableRawBufferPointer) -> Int {
+    // Not implemented
+    return 0
```

Should we throw an error here instead?

---

Unfortunately our only other option is to `abort()` here because this function signature is required by the `ArchiveByteStreamProtocol` - i.e. we can't add a `throws`.

In practice, I've found that the framework ends the compression process abruptly when no bytes are read or written, so I'm not sure if there's a better approach here :-(.